# Assignment 1 - Build a performing machine Regression algorithm

### Objective:
The objective of the assignment is to develop a linear regression machine learning model

### Data Description:
We have collected data on students' performance from UCI data respository. The data shows peformance of students in Gabriel Pereira and Mousinho da Silveira schools. The dataset has the following features:

### Attributes of the dataset that shows students performance in math course 
1. school - student's school (binary: 'GP' - Gabriel Pereira or 'MS' - Mousinho da Silveira) 
2. sex - student's sex (binary: 'F' - female or 'M' - male) 
3. age - student's age (numeric: from 15 to 22) 
4. address - student's home address type (binary: 'U' - urban or 'R' - rural) 
5. famsize - family size (binary: 'LE3' - less or equal to 3 or 'GT3' - greater than 3) 
6. Pstatus - parent's cohabitation status (binary: 'T' - living together or 'A' - apart) 
7. Medu - mother's education (numeric: 0 - none, 1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education) 
8. Fedu - father's education (numeric: 0 - none, 1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education) 
9. Mjob - mother's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other') 
10. Fjob - father's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other') 
11. reason - reason to choose this school (nominal: close to 'home', school 'reputation', 'course' preference or 'other') 
12. guardian - student's guardian (nominal: 'mother', 'father' or 'other') 
13. traveltime - home to school travel time (numeric: 1 - <15 min., 2 - 15 to 30 min., 3 - 30 min. to 1 hour, or 4 - >1 hour) 
14. studytime - weekly study time (numeric: 1 - <2 hours, 2 - 2 to 5 hours, 3 - 5 to 10 hours, or 4 - >10 hours) 
15. failures - number of past class failures (numeric: n if 1<=n<3, else 4) 
16. schoolsup - extra educational support (binary: yes or no) 
17. famsup - family educational support (binary: yes or no) 
18. paid - extra paid classes within the course subject (Math or Portuguese) (binary: yes or no) 
19. activities - extra-curricular activities (binary: yes or no) 
20. nursery - attended nursery school (binary: yes or no) 
21. higher - wants to take higher education (binary: yes or no) 
22. internet - Internet access at home (binary: yes or no) 
23. romantic - with a romantic relationship (binary: yes or no) 
24. famrel - quality of family relationships (numeric: from 1 - very bad to 5 - excellent) 
25. freetime - free time after school (numeric: from 1 - very low to 5 - very high) 
26. goout - going out with friends (numeric: from 1 - very low to 5 - very high) 
27. Dalc - workday alcohol consumption (numeric: from 1 - very low to 5 - very high) 
28. Walc - weekend alcohol consumption (numeric: from 1 - very low to 5 - very high) 
29. health - current health status (numeric: from 1 - very bad to 5 - very good) 
30. absences - number of school absences (numeric: from 0 to 93) 

### These are grades secured by students at the end of each period
31. G1 - first period grade (numeric: from 0 to 20) 
31. G2 - second period grade (numeric: from 0 to 20) 
32. G3 - final grade (numeric: from 0 to 20, output target)

In [1]:
# import essential libraries
import pandas as pd
import numpy as np
pd.set_option("display.max_columns", None)
# adjust screensize
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

/var/folders/qm/tty6fsf56cb0rsb2r8l0_6cc0000gn/T/ipykernel_60832/922315407.py:6: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML


In [2]:
mdata = pd.read_csv("student-mat.csv", sep=";")
mdata.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  
0     1       3         6   5   6   6  
1     1       3         4   5   5   6  
2     3       3        10   7   8  10  
3     1       5         2  15  14  15  
4     2       5         4   6  10  10

In [3]:
mdata.info() # identifying the categorical varibles and checking for null values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

In [4]:
mdata.shape

(395, 33)

As one can see above, there is a total of 395 samples of data with 33 features. Fortunately, none of the features has any missing values, so there is no need to either drop missing values or to impute them. However, we notice that many features are stored as objects (categorical values). As a result, we need to convert these values to numbers so that the model can interpret them.

In [5]:
mdata.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  
0     1       3         6   5   6   6  
1     1       3         4   5   5   6  
2     3       3        10   7   8  10  
3     1       5         2  15  14  15  
4     2       5         4   6  10  10

In [6]:
for i in mdata.columns:
    print(f"Values in {i}: {mdata[i].unique()}")

Values in school: ['GP' 'MS']
Values in sex: ['F' 'M']
Values in age: [18 17 15 16 19 22 20 21]
Values in address: ['U' 'R']
Values in famsize: ['GT3' 'LE3']
Values in Pstatus: ['A' 'T']
Values in Medu: [4 1 3 2 0]
Values in Fedu: [4 1 2 3 0]
Values in Mjob: ['at_home' 'health' 'other' 'services' 'teacher']
Values in Fjob: ['teacher' 'other' 'services' 'health' 'at_home']
Values in reason: ['course' 'other' 'home' 'reputation']
Values in guardian: ['mother' 'father' 'other']
Values in traveltime: [2 1 3 4]
Values in studytime: [2 3 1 4]
Values in failures: [0 3 2 1]
Values in schoolsup: ['yes' 'no']
Values in famsup: ['no' 'yes']
Values in paid: ['no' 'yes']
Values in activities: ['no' 'yes']
Values in nursery: ['yes' 'no']
Values in higher: ['yes' 'no']
Values in internet: ['no' 'yes']
Values in romantic: ['no' 'yes']
Values in famrel: [4 5 3 1 2]
Values in freetime: [3 2 4 1 5]
Values in goout: [4 3 2 1 5]
Values in Dalc: [1 2 5 3 4]
Values in Walc: [1 3 2 4 5]
Values in health: [3 5

In [7]:
mdata.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  395.000000  395.000000  395.000000  395.000000  
mean     5.708861   10.908861   10.713924   10.415190  
std      8.003096    3.319195    3.761505    4.581443  
min      0.000000    3.000000    0.000000    0.000000  
25%      0.000000    8.000000    9.000000    8.000000  
50%      4.000000   11.000000   11.000000   11.000000  
75%      8.000000   13.000000   13.000000   14.000000  
max     75.000000   19.000000   19.000000   20.000000

In [8]:
print("The dataset about Mathematics has %d rows and %d columns " % mdata.shape + "and has " + ("some" if mdata.isnull().values.any() else "no") + " missing values.")

The dataset about Mathematics has 395 rows and 33 columns and has no missing values.


The above view of data enables us to determine which features need to be translated into dummy variables. 

In [9]:
gradesMat = mdata["G3"].value_counts().sort_index()
#print(gradesMat.values)
#print(gradesMat.index)

yMat = gradesMat.values
xMat =[]
for n in gradesMat.index:
  xMat.append(str(n))


def valueToSwitch(n):
    if(n<=9):
      return 0
    if(n==10 or n==11):
      return 1
    if(n==12 or n==13):
      return 2
    if(n==14 or n==15):
      return 3
    else:
      return 4


fiveLevelGradesMat = [0,0,0,0,0]

for grade, quantity in gradesMat.items():
    level = valueToSwitch(grade)
    fiveLevelGradesMat[level] = fiveLevelGradesMat[level] + quantity



In [10]:
pip install plotly

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.2/15.2 MB 14.6 MB/s eta 0:00:0000:0100:01
Note: you may need to restart the kernel to use updated packages.


In [11]:
import plotly
import plotly.graph_objs as go
import plotly.express as px
import plotly.figure_factory as ff
from plotly.offline import init_notebook_mode
init_notebook_mode(connected=True) # to show plots in notebook

In [12]:
colors = plotly.colors.DEFAULT_PLOTLY_COLORS

binaryGradesMat = [fiveLevelGradesMat[0], fiveLevelGradesMat[1]+fiveLevelGradesMat[2]+fiveLevelGradesMat[3]+fiveLevelGradesMat[4]]

# Create figure
fig = go.Figure()

fiveLevel = ['0-9', '10-11', '12-13', '14-15', '16-20']
twoLevel = ['Fail', 'Pass']

colors20L = [colors[2]]*len(xMat)
for i in list(range(0,7)):
    colors20L[i] = colors[3]

    
colors5L = [colors[2]]*5
colors5L[0] = colors[3]

colors2L = [colors[3],colors[2]]

fig.add_trace(
        go.Bar(
            visible=False,
            marker_color=colors20L,
            name='20Level Mat',
            x=xMat,
            y=yMat,
            text=yMat,
            textposition='auto'))

fig.add_trace(
        go.Bar(
            visible=False,
             marker_color=colors5L,
            name='5Level Mat',
            x=list(range(1,6)),
            y=fiveLevelGradesMat,
            text=fiveLevelGradesMat,
            textposition='auto'))

fig.add_trace(
        go.Bar(
            visible=False,
            marker_color=colors2L,
            name='2Level Mat',
            x=list(range(1,3)),
            y=binaryGradesMat,
            text=binaryGradesMat,
            textposition='auto'))



fig.data[0].visible = True

# Create and add slider
steps = []
labels = ['20 Level Mat', '5 Level Mat', '2 Level Mat', '20 Level Por', '5 Level Por', '2 Level Por']
for i in range(len(fig.data)):
    step = dict(
        method="restyle",
        args=["visible", [False] * len(fig.data)],
        label = labels[i]
    )
    step["args"][1][i] = True  # Toggle i'th trace to "visible"
    steps.append(step)

sliders = [dict(
    active=0,
    currentvalue={"prefix": "Grades: "},
    pad={"t": 50},
    steps=steps
)]

fig.update_layout(
    xaxis = dict(
        tickmode='linear'),
    sliders=sliders,
    barmode='group',
    title='Histogran of final scores in Mathematics',
    xaxis_title="Grade",
    yaxis_title="#student"
)

fig.show()

In [13]:
#CAPIRE SE ALCUNE CATEGORIE BINARIE VANNO CODIFICATE COME 1/0 OPPURE COME DUMMY

#cleanup_nums = {"yes": 1, "no": 0}
cleanup_nums = {"yes": 1, "no": 0, "F": 1, "M": 0, "GP": 1, "MS": 0, "R": 1, "U" : 0, "GT3": 1, "LE3" :0, "A":1, "T":0}

mdata.replace(cleanup_nums, inplace=True)

#dataMatExt= pd.get_dummies(dataMat, columns=["school", "sex", "address", "famsize", "Pstatus", "Mjob", "Fjob", "reason", "guardian"])
#dataPorExt= pd.get_dummies(dataPor, columns=["school", "sex", "address", "famsize", "Pstatus", "Mjob", "Fjob", "reason", "guardian"])

dataMatExt= pd.get_dummies(mdata, columns=[ "Mjob", "Fjob", "reason", "guardian"])

#print(dataMatExt.columns)

#G1, G2, e G3 li faccio diventare binari così da semplificare e capire meglio quali attributi siano collegati a una promozione o bocciatura
def gradesToBinary(x):
  if(x>9):
    return 1
  else:
    return 0

fiveLevel = ['0-9', '10-11', '12-13', '14-15', '16-20']

def gradesTo5L(x):
    if(x<10):
        return 0
    elif (x==10 or x==11):
        return 1
    elif (x==12 or x==13):
        return 2
    elif (x==14 or x==15):
        return 3
    else:
        return 4
    

dataMat20L = dataMatExt.copy()

dataMat5L = dataMatExt.copy()

dataMat5L['G3'] = dataMat5L['G3'].map(gradesTo5L)
dataMat5L['G2'] = dataMat5L['G2'].map(gradesTo5L)
dataMat5L['G1'] = dataMat5L['G1'].map(gradesTo5L)

dataMatExt['G1'] = dataMatExt['G1'].map(gradesToBinary)
dataMatExt['G2'] = dataMatExt['G2'].map(gradesToBinary)
dataMatExt['G3'] = dataMatExt['G3'].map(gradesToBinary)


#calculate correlation matrix
cor1 =dataMatExt.corr()


def get_redundant_pairs(df):
    '''Get diagonal and lower triangular pairs of correlation matrix'''
    pairs_to_drop = set()
    cols = df.columns
    for i in range(0, df.shape[1]):
        for j in range(0, i+1):
            pairs_to_drop.add((cols[i], cols[j]))
    return pairs_to_drop


#take the lower trinagular pairs
correlationCouple = cor1.unstack().drop(labels=get_redundant_pairs(cor1))

#Stampa le coppie maggiormente correlate
#cc = correlationCouple.sort_values(ascending=False)
#print(cc)

#for x,y in zip(cc.index, cc):
#    print(x,y)

#take only the value associated with G3
correlationCoupleG3 = correlationCouple.xs('G3', level=0, drop_level=False).append(correlationCouple.xs('G3', level=1, drop_level=False)).sort_values(ascending=False)
 

def top_correlated(data, s):
    i = 0
    topIndex =  []
    topValue = []
    if(s=="M"):
        value = 0.09
    else:
        value = 0.1
    
    for x in data:
        if(x>value or x< -value):
          #print(x)
          tmp1 =x
          tmp2 = data.index[i]
          topIndex.append(tmp2)
          topValue.append(tmp1)
        i = i +1

    top = dict(zip(topValue, topIndex))
    return top
   

#topCorrelated = correlationCoupleG3.iloc[:5].append(correlationCoupleG3.iloc[correlationCoupleG3.size-5:])

#Coppie con correlazione > |0.09|
topCorrelated = top_correlated(correlationCoupleG3, "M")

topValuePos= []
topValueNeg= []
topIndexPos=[]
topIndexNeg = []

for x, y in topCorrelated.items():
  if(x>0):
    topValuePos.append(x)
    if(y[0]=='G3'):
      topIndexPos.append(y[1])
    else:
      topIndexPos.append(y[0])
  else:
    topValueNeg.append(x) 
    if(y[0]=='G3'):
      topIndexNeg.append(y[1])
    else:
      topIndexNeg.append(y[0])
    
#Non rappresentiamo G1 e G2 perchè la loro correlazione è evidente e non permettono una buona rappresentazione degli altri elementi


fig = go.Figure()

fig.add_trace(go.Bar(x=topIndexPos[2:], y=topValuePos[2:],
                base=0,
                marker_color=colors[9],
                name='positive correlation Mat',
                hoverinfo='text',
                hovertext = [(x,y) for (x,y) in zip(topIndexPos[2:], topValuePos[2:])]
                ))

fig.add_trace(go.Bar(x=topIndexNeg, y=[abs(ele) for ele in topValueNeg],
                base=topValueNeg,
                marker_color= colors[8],
                name='negative correlation Mat',
                hoverinfo='text',
                hovertext =[(x,y) for (x,y) in zip(topIndexNeg, topValueNeg)]))


#fig.data[2].visible = False
#fig.data[3].visible = False

stepsG3 = []
labels = ['Mathematic', 'Mathematic']
for i in range(len(fig.data)):
    if(i==0 or i==2):
        step = dict(
        method="restyle",
        args=["visible", [False] * len(fig.data)],
        label = labels[i]
        )
        step["args"][1][i] = True  # Toggle i'th trace to "visible"
        step["args"][1][i+1] = True  # Toggle i'th trace to "visible"
        stepsG3.append(step)
    

sliders = [dict(
    active=0,
    currentvalue={"prefix": "Subject:   "},
    pad={"t": 70},
    steps=stepsG3
)]

fig.update_layout(
    sliders=sliders,
    barmode='group',
    title='Correlation of key attributes to Final Grade in Mathematics',
    xaxis_title="Attribute",
    yaxis_title="Correlation with G3"
)

fig.show()

/var/folders/qm/tty6fsf56cb0rsb2r8l0_6cc0000gn/T/ipykernel_60832/3248758116.py:75: FutureWarning:

The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.



In [14]:
# a function created to convert categorical features into dummy variables.
def dummy_creator(arr, pre):
    arr_dummy = pd.get_dummies(arr, prefix=pre, drop_first=True) # dropped the first dummy variable to avoid dummy variable trap
    return arr_dummy

In [15]:
mdata_categorical = mdata.drop(mdata.select_dtypes([np.number]), axis=1) # to extract all categorical data

In [16]:
for col in mdata_categorical.columns:
    mdata_categorical = pd.concat([mdata_categorical, dummy_creator(mdata_categorical[col], col)], axis=1) # converts features into dummy variables and concatenate them to the dataframe
mdata_categorical.head()

Mjob      Fjob  reason guardian  Mjob_health  Mjob_other  Mjob_services  \
0  at_home   teacher  course   mother            0           0              0   
1  at_home     other  course   father            0           0              0   
2  at_home     other   other   mother            0           0              0   
3   health  services    home   mother            1           0              0   
4    other     other    home   father            0           1              0   

   Mjob_teacher  Fjob_health  Fjob_other  Fjob_services  Fjob_teacher  \
0             0            0           0              0             1   
1             0            0           1              0             0   
2             0            0           1              0             0   
3             0            0           0              1             0   
4             0            0           1              0             0   

   reason_home  reason_other  reason_reputation  guardian_mother  \
0            0             0                  0                1   
1            0             0                  0                0   
2            0             1                  0                1   
3            1             0                  0                1   
4            1             0                  0                0   

   guardian_other  
0               0  
1               0  
2               0  
3               0  
4               0

We notice above that all the categorical features have been converted to dummy values and concatenated to the categorical dataset

In [17]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler() # importing MinMax Scaler to scale numerical data within the range of [0,1] so that they are consistent with the categorical ones. This will ensure 
# that data with greater magnitude will not sway the model results towards itself. 

In [18]:
mdata_continuous = mdata._get_numeric_data() # extracting all the numeric data from the original dataset
mdata_cont_scaled = pd.DataFrame(scaler.fit_transform(mdata_continuous), columns=mdata_continuous.columns) # scaled the continuous variables so that they are in the 
# same range as the categoricla ones
mdata_cont_scaled.head()

school  sex       age  address  famsize  Pstatus  Medu  Fedu  traveltime  \
0     1.0  1.0  0.428571      0.0      1.0      1.0  1.00  1.00    0.333333   
1     1.0  1.0  0.285714      0.0      1.0      0.0  0.25  0.25    0.000000   
2     1.0  1.0  0.000000      0.0      0.0      0.0  0.25  0.25    0.000000   
3     1.0  1.0  0.000000      0.0      1.0      0.0  1.00  0.50    0.000000   
4     1.0  1.0  0.142857      0.0      1.0      0.0  0.75  0.75    0.000000   

   studytime  failures  schoolsup  famsup  paid  activities  nursery  higher  \
0   0.333333       0.0        1.0     0.0   0.0         0.0      1.0     1.0   
1   0.333333       0.0        0.0     1.0   0.0         0.0      0.0     1.0   
2   0.333333       1.0        1.0     0.0   1.0         0.0      1.0     1.0   
3   0.666667       0.0        0.0     1.0   1.0         1.0      1.0     1.0   
4   0.333333       0.0        0.0     1.0   1.0         0.0      1.0     1.0   

   internet  romantic  famrel  freetime  goout  Dalc  Walc  health  absences  \
0       0.0       0.0    0.75      0.50   0.75  0.00  0.00     0.5  0.080000   
1       1.0       0.0    1.00      0.50   0.50  0.00  0.00     0.5  0.053333   
2       1.0       0.0    0.75      0.50   0.25  0.25  0.50     0.5  0.133333   
3       1.0       1.0    0.50      0.25   0.25  0.00  0.00     1.0  0.026667   
4       0.0       0.0    0.75      0.50   0.25  0.00  0.25     1.0  0.053333   

       G1        G2    G3  
0  0.1250  0.315789  0.30  
1  0.1250  0.263158  0.30  
2  0.2500  0.421053  0.50  
3  0.7500  0.736842  0.75  
4  0.1875  0.526316  0.50

In [19]:
mdata_new = pd.concat([mdata_categorical, mdata_cont_scaled], axis=1) # merged the numerical data with categorical ones
mdata_new = mdata_new._get_numeric_data() # since categorical data has been converted to dummy variables, we now remove the original categorical data with string values
mdata_new

Mjob_health  Mjob_other  Mjob_services  Mjob_teacher  Fjob_health  \
0              0           0              0             0            0   
1              0           0              0             0            0   
2              0           0              0             0            0   
3              1           0              0             0            0   
4              0           1              0             0            0   
..           ...         ...            ...           ...          ...   
390            0           0              1             0            0   
391            0           0              1             0            0   
392            0           1              0             0            0   
393            0           0              1             0            0   
394            0           1              0             0            0   

     Fjob_other  Fjob_services  Fjob_teacher  reason_home  reason_other  \
0             0              0             1            0             0   
1             1              0             0            0             0   
2             1              0             0            0             1   
3             0              1             0            1             0   
4             1              0             0            1             0   
..          ...            ...           ...          ...           ...   
390           0              1             0            0             0   
391           0              1             0            0             0   
392           1              0             0            0             0   
393           1              0             0            0             0   
394           0              0             0            0             0   

     reason_reputation  guardian_mother  guardian_other  school  sex  \
0                    0                1               0     1.0  1.0   
1                    0                0               0     1.0  1.0   
2                    0                1               0     1.0  1.0   
3                    0                1               0     1.0  1.0   
4                    0                0               0     1.0  1.0   
..                 ...              ...             ...     ...  ...   
390                  0                0               1     0.0  0.0   
391                  0                1               0     0.0  0.0   
392                  0                0               1     0.0  0.0   
393                  0                1               0     0.0  0.0   
394                  0                0               0     0.0  0.0   

          age  address  famsize  Pstatus  Medu  Fedu  traveltime  studytime  \
0    0.428571      0.0      1.0      1.0  1.00  1.00    0.333333   0.333333   
1    0.285714      0.0      1.0      0.0  0.25  0.25    0.000000   0.333333   
2    0.000000      0.0      0.0      0.0  0.25  0.25    0.000000   0.333333   
3    0.000000      0.0      1.0      0.0  1.00  0.50    0.000000   0.666667   
4    0.142857      0.0      1.0      0.0  0.75  0.75    0.000000   0.333333   
..        ...      ...      ...      ...   ...   ...         ...        ...   
390  0.714286      0.0      0.0      1.0  0.50  0.50    0.000000   0.333333   
391  0.285714      0.0      0.0      0.0  0.75  0.25    0.333333   0.000000   
392  0.857143      1.0      1.0      0.0  0.25  0.25    0.000000   0.000000   
393  0.428571      1.0      0.0      0.0  0.75  0.50    0.666667   0.000000   
394  0.571429      0.0      0.0      0.0  0.25  0.25    0.000000   0.000000   

     failures  schoolsup  famsup  paid  activities  nursery  higher  internet  \
0    0.000000        1.0     0.0   0.0         0.0      1.0     1.0       0.0   
1    0.000000        0.0     1.0   0.0         0.0      0.0     1.0       1.0   
2    1.000000        1.0     0.0   1.0         0.0      1.0     1.0       1.0   
3    0.000000        0.0     1.0   1.0         1.0      1.0     1.0       1.0

In [20]:
# We split the dataset into dependent (Final grade-G3) and independent variables (all the remaining features)
# Given our lack of domain expertise, we have not eliminated any feature yet. We intend to do that later in the analysis based on their effect on the final score.
x = mdata_new.iloc[:, :-1]
y = mdata_new.iloc[:, -1]

In [21]:
# import train test split package from scikit learn library
from sklearn.model_selection import train_test_split

In [22]:
# We are dividing the dataset into training and testing splits. We retain 20% of the data for testing purpose
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [23]:
from sklearn.linear_model import LinearRegression

In [24]:
model = LinearRegression() # creating a linear regression object

In [25]:
model.fit(x_train, y_train)

LinearRegression()

In [26]:
from sklearn.metrics import r2_score

In [27]:
y_predicted = model.predict(x_test)

In [28]:
score_wo_RFE = r2_score(y_predicted, y_test) # this score is based on all the features prior to applying any form of feature selection
score_wo_RFE

0.7102320490925694

In [29]:
from sklearn.feature_selection import RFE # importing Recursive Feature Elimination feature from Scikit Learn Library

In [30]:
feature_num = list()
scores = list()
for index in range(1,41):
    sel = RFE(LinearRegression(), n_features_to_select=index) # we use a loop to iterate through the number of features starting from 1 to 41 (the total number of features)
    sel.fit(x_train, y_train)
    x_train_rfe = sel.transform(x_train) # RFE is applied on training set based on the number of features selected iteratively
    x_test_rfe = sel.transform(x_test) # RFE is applied on test data based on the number of features selected iteratively
    model = LinearRegression() 
    model.fit(x_train_rfe, y_train)
    y_predicted = model.predict(x_test_rfe)
    feature_num.append(index)
    scores.append(r2_score(y_predicted, y_test)) # predicting r2 scores for combination of features from 1 to 41

score_dict = dict(zip(feature_num, scores))
score_dict # a dictionary that shows number of features as keys and accuracy scores as values

{1: 0.7595167048116915,
 2: 0.7556024757151003,
 3: 0.7704297228788032,
 4: 0.7731287745087871,
 5: 0.7745936863734233,
 6: 0.7629448118384674,
 7: 0.7521890406006199,
 8: 0.7452886904816267,
 9: 0.7386850045269885,
 10: 0.7298416554465672,
 11: 0.7243661440810731,
 12: 0.7217916106222886,
 13: 0.7223742626227205,
 14: 0.7268821319734144,
 15: 0.7263938877853677,
 16: 0.7274509119154693,
 17: 0.7196589420849067,
 18: 0.7203266271850113,
 19: 0.7184690754794401,
 20: 0.7165867624916797,
 21: 0.714766802011982,
 22: 0.7147689339529151,
 23: 0.7109168365676874,
 24: 0.7121118207064285,
 25: 0.7091867745472342,
 26: 0.7090021270929521,
 27: 0.709173827454372,
 28: 0.7094938997206494,
 29: 0.7103012794527541,
 30: 0.7110552547495368,
 31: 0.7100133214229578,
 32: 0.7095480296900621,
 33: 0.7101530020211748,
 34: 0.711321693920824,
 35: 0.7123262633526641,
 36: 0.7115052456301367,
 37: 0.7111562254275205,
 38: 0.7102886030751941,
 39: 0.7100218246676412,
 40: 0.7102845072992838}

In [31]:
max_key = max(score_dict, key=score_dict.get)
max_key # show that RFE has created a combination of five features that produces the maximum accuracy score

5

### Understanding the Features and their implications
Since we have already discovered that RFE provides the highest scores when five features are selected, we run the model with 5 features and note that results below.

In [32]:
select = RFE(LinearRegression(), n_features_to_select=5)

In [33]:
select.fit(x_train, y_train)
x_train_selected = select.transform(x_train)
x_test_selected = select.transform(x_test)

In [34]:
model_new = LinearRegression()

In [35]:
model_new.fit(x_train_selected, y_train)

LinearRegression()

In [36]:
y_predicted_new = model_new.predict(x_test_selected)

In [37]:
score_with_RFE = r2_score(y_predicted_new, y_test)
score_with_RFE

0.7745936863734233

In [38]:
x_train.columns[select.get_support()]
# Based on RFE we discover that age, quality of family relationships, number of absenses, scores in first and second periods are the most influencial features when predicting 
# final scores of students in mathematics.

Index(['age', 'famrel', 'absences', 'G1', 'G2'], dtype='object')

In [39]:
# change in score
"{:.2f}%".format((score_with_RFE - score_wo_RFE)/score_wo_RFE * 100)

'9.06%'

### Conclusion:
We set out to create a machine learning linear regression model that predicts final score of students in mathematics. We started with 32 features and got an R2 score of 0.71. Given any R2 score above 0.6 signifies a good model, it is safe to assume that our model is predicting scores well. However, we wanted to enhance the effectiveness and efficiency of the model by recursively eliminating features and assessing the impact on the accuracy score. 

Using the Recursive Feature Elimination package of Scikit Learn Library, we discovered that the best R2 score is achieved when we limit the number of features to five (age, quality of family relationship, number of absences, score in first period and score in second period). By using RFE, we managed to increase the accuracy by more than 9 percentage points and reduce the number of features from 32 to 5. In addition to providing more reliable predictions, we believe application of RFE will significantly reduce strain on computational resources.